### DISCUSSION TITLE
Adjust velocity mid-jump?

### DISCUSSION THREAD

**Discussion by joshuacurtiss** [replit/kaboom#779] [Q&A]
Hello! Playing with v3000.

You know how some games including Mario have the classic feature where pushing jump will do a normal jump but if you push and hold on the jump button you'll get a slightly bigger jump. I'm trying to implement that in Kaboom and having trouble.  I figured I could have the jump force be for a normal jump but if the user continues to push the jump button, keep on adding a little extra to velocity y while they're still actively jumping.

A crude example:

```js
player.onUpdate(() => {
   if (player.isJumping() && isKeyDown('space')) {
      player.vel.y+=5; // But you can't do this
   }
}
```

Another situation is if you wanted to have the behavior where the player stomps down while in mid-air, effectively changing the upward velocity to suddenly be a downward velocity. 

Is there any way to handle this in v3000?

Josh

**Comment by luluxia** [comment]
kaboom itself doesn't seem to have this functionality, so I implemented it by modifying the kaboom source code and adding the following function to the [body component](https://github.com/replit/kaboom/blob/master/src/kaboom.ts#L4153C50-L4153C50).
```
jumpRelease() {
  if (vel.y < 0) {
    vel.y *= 0.5
  }
}
```
When the jump button is released, trigger this function to end the jump early.

Regarding the other scenario, which I haven't implemented yet, it might be possible to use a similar approach.

**Reply by joshuacurtiss** [reply]
Thanks. Yeah, that's my finding too. I like your approach, by the way. 

If I have to change the core library code without being able to use a plugin or hook mechanism, that makes me sad. :( 

By the way, I took a combination of our ideas and did a pull request. Dear replit folks, please have a look... 😄 
https://github.com/replit/kaboom/pull/781

Josh

**Reply by joshuacurtiss** [reply]
This update in 3000.1.17 🎉
https://github.com/replit/kaboom/blob/925eb137/CHANGELOG.md#v3000117
